lib/errors/page_content_errors.rb

```
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+module Errors
+  module PageContentErrors
+    class NilPageContentError < StandardError
```

I think this may be a confusing way of naming and handling this error.

We have a pretty good idea that this error is specific to WikiApi when it attempts to fetch page content and fails to do so, and it's not a 404.

After looking over the different places in the code that use WikiApi#get_page_content, I think none of them are written in a way that explicitly handles this kind of failure adequately, as the current behavior generally treats an empty string (returned from a 404) the same way as `nil`, so we may be missing other problems that we don't know about.

My current thinking is that we should actually raise an error from WikiApi rather than return `nil`, and then figure out the next steps from there.

---

```
 # Returns nil if it cannot get any info from the wiki, but returns
  # empty string if it's a 404 because the page is a redlink.
  def get_page_content(page_title)
    response = mediawiki('get_wikitext', page_title)
    case response&.status
    when 200
      response.body.force_encoding('UTF-8')
    when 404
      ''
    end
  end
```
Should i add a case for the error being 429 and let it raise an error from there, or a general else case and raise the error from there? or both?
 I could create an error class specific to 429 errors since that's what we assume is happening and raise it when the status is 429, then raise the more general NilPageContentError (renamed) for the else case.
 
 In that case would the nil guards still be needed in the code that use WikiApi#get_page_content, since when errors are raised, I assume the entire method execution will immediately stop.

---

I think we could just throw the errors from here, and decide later whether to add error handling at any of the specific places in the code that are affected by it.

The trick here is that there are some 429 errors that we are intentionally basically ignoring, but not ones that use .get_page_content. Ideally, we'd be raising those errors based on status within the method that actually makes the call, so that would wouldn't need parallel handling at every method that share `mediawiki`. But 429 is the only kind of error that `mediawiki` already handles, so I don't think you'll be able to detect when it's a 429 here unless you modify `mediawiki`, for example, by adding an optional parameter for raising an error rather than returning nil. I guess that's probably the right strategy here.

---

Got it, I'll revert the changes except the custom error class and add this to wiki_api:
```
else
      exception = Errors::PageContentErrors::NilPageContentError.new(page_title, response&.status)
      raise exception
    end
```
so the new message will be: `"Failed to fetch content for #{page} with response status: #{status}."`

For the renaming of the error, I can as well remove the module and just have a direct class like `Errors::PageFetchError`. I'm not too clear on what a good name is though, so your suggestions are appreciated.

About `mediawiki`, I took a look at the code in the separate repo and I ended up confusing myself a bit as to why 429 errors would occur here given that other 429 instances usually log a `MediawikiApi::HttpError: unexpected HTTP response (429)` at this line : `@mediawiki.send(action, query)` so maybe it's other error codes causing this?

---

Hmm... I'm not sure.

For the error, since this is a type of error we expect to only originate from WikiApi, my suggestion would be to define it within WikiApi rather than in the separate module.

---

Created a new PR and branch for it: https://github.com/WikiEducationFoundation/WikiEduDashboard/pull/6158